File: app/src/main/java/com/venditti/placares/MainActivity.java
package com.venditti.placares;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.venditti.placares.activities.BiscaActivity;
import com.venditti.placares.adapter.PlayerAdapter;
import com.venditti.placares.config.ConfigFireBase;
import com.venditti.placares.databinding.ActivityMainBinding;
import com.venditti.placares.helper.DateCustom;
import com.venditti.placares.model.BiscaViewModel;
import com.venditti.placares.model.Players;

import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private PlayerAdapter adapter;
    private String game;
    private BiscaViewModel viewModel;
    AtomicInteger gameCount = new AtomicInteger(0);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        //configuração toolbar
        Toolbar toolbar = binding.toolbarContext.toolbar;
        toolbar.setTitle("Placares");
        setSupportActionBar(toolbar);

        carregaJogosPassados();

        game = binding.game.getSelectedItem().toString();
        Spinner spinner = binding.numberPlayers;
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent, View view, int pos, long id) {
                int qtdJogadores = pos+3;

                carregarRecycleView(qtdJogadores);
            }
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        binding.btnPlay.setOnClickListener(v -> {
            startGame();
        });
    }

    private void startGame() {
        if(adapter.getMapPlayers().isEmpty()){
            Toast.makeText(getApplicationContext(), "preencher nome dos jogadores", Toast.LENGTH_SHORT).show();
        }else {
            Map<Integer, Players> mapPlayers = adapter.getMapPlayers();
            ArrayList<Players> players = new ArrayList<>(mapPlayers.values());

            players.forEach(p -> p.salvar(game, gameCount.toString()));

            Intent intent = new Intent(getApplicationContext(), BiscaActivity.class);
            startActivity(intent);
        }
    }

    private void carregaJogosPassados() {
        ConfigFireBase.getDataGameBiscaReference().addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                snapshot.getChildren().forEach(s -> {
                    gameCount.getAndIncrement();
                });
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    private void carregarRecycleView(int count){
        RecyclerView rv = binding.recyclerPlayer;
        //configurar Adapter
        adapter = new PlayerAdapter(count);

        //configurar RecyclerView
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getApplicationContext());
        rv.setLayoutManager(layoutManager);
        rv.setHasFixedSize(false);
        rv.setAdapter(adapter);
    }
}

File: app/src/main/java/com/venditti/placares/MainActivity.kt
package com.venditti.placares

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.venditti.placares.activities.BiscaActivity
import com.venditti.placares.adapter.PlayerAdapter
import com.venditti.placares.config.ConfigFireBase
import com.venditti.placares.databinding.ActivityMainBinding
import com.venditti.placares.model.Players
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer

class MainActivity : AppCompatActivity() {
    private var binding: ActivityMainBinding? = null
    private var adapter: PlayerAdapter? = null
    private var game: String? = null
    private var gameCount: AtomicInteger? = null
    private var qtdPlayers: Int? = null

    override fun onStart() {
        super.onStart()
        gameCount = AtomicInteger(0)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding!!.root)

        //configuração toolbar
        val toolbar = binding!!.toolbarContext.toolbar
        toolbar.title = "Placares"
        setSupportActionBar(toolbar)

        carregaJogosPassados()

        game = binding!!.game.selectedItem.toString()
        val spinner = binding!!.numberPlayers
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, pos: Int, id: Long) {
                qtdPlayers = pos + 3
                carregarRecycleView(qtdPlayers!!)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
        binding!!.btnPlay.setOnClickListener { startGame() }
    }

    private fun startGame() {
        val mapPlayers = adapter!!.getMapPlayers()
        val players = ArrayList(mapPlayers!!.values)
        if (adapter!!.getMapPlayers().size > qtdPlayers!!) {
            Toast.makeText(applicationContext, "preencher nome dos jogadores", Toast.LENGTH_SHORT).show()
        } else {
            players.forEach(Consumer { p: Players? -> p!!.salvar(game, gameCount.toString()) })
        }
        val intent = Intent(applicationContext, BiscaActivity::class.java)
        intent.putExtra("listaJogadores", players)
        startActivity(intent)
    }

    private fun carregaJogosPassados() {
        ConfigFireBase.dataGameBiscaReference.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                snapshot.children.forEach(Consumer { s: DataSnapshot? -> gameCount!!.getAndIncrement() })
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun carregarRecycleView(count: Int) {
        val rv = binding!!.recyclerPlayer
        //configurar Adapter
        adapter = PlayerAdapter(count)

        //configurar RecyclerView
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(applicationContext)
        rv.layoutManager = layoutManager
        rv.setHasFixedSize(false)
        rv.adapter = adapter
    }
}